[16.0][MIG] openupgrade_framework, base v16

hbrunn opened the pull request:
this is preliminary because I didn't yet go through all monkey patches to verify we still need them, but this lets the base migration run without crashing

TODO:
- [x] verify we still need all monkey patches and no new ones
- [x] re-add the testing framework for migration scripts

legalsylvain committed:
[INIT] Full refactor of the OpenUpgrade project

StefanRijnhart committed:
[DEL] openupgrade_records -> server-tools/14.0

StefanRijnhart committed:
[RFR] OpenUpgrade framework patches

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[ADD] icon.png

MiquelRForgeFlow committed:
[FIX] _logger.warn -> _logger.warning

MiquelRForgeFlow committed:
[FIX] doc: base must be included in --load parameter

If we don't include it, /jsonrpc route doesn't work.

OCA-git-bot committed:
[UPD] README.rst

MiquelRForgeFlow committed:
[FIX] base: Don't fail uninstalling when missing models

If you have a module in previous versions that adds data on a model,
and such model is not loaded in the registry in current version because
the module is absent in it, you can't uninstall such module, getting
this error:

  File "odoo/odoo/addons/base/models/ir_model.py", line 1945, in _module_data_uninstall
    delete(self.env[model].browse(item[1] for item in items))
  File "odoo/odoo/api.py", line 463, in __getitem__
    return self.registry[model_name]._browse(self, (), ())
  File "odoo/odoo/modules/registry.py", line 177, in __getitem__
    return self.models[model_name]
KeyError: 'model'

With this patch, data cleanup of such model is skipped and there's no crash.

MiquelRForgeFlow committed:
[FIX] openupgrade_framework: _drop_column was in wrong class

legalsylvain committed:
[DOC] add banner file, to make this module more visible on the app store ; add links to the new OpenUpgrade website

OCA-git-bot committed:
[UPD] README.rst

StefanRijnhart committed:
[FIX] Don't break on missing model in selection _process_ondelete

legalsylvain committed:
[REF] Add maintainers

OCA-git-bot committed:
[UPD] README.rst

MiquelRForgeFlow committed:
[IMP] openupgrade_framework: avoid AttributeError in view validation

If corresponding field is None, we need to avoid the "AttributeError: 'NoneType' object has no attribute error.

legalsylvain committed:
[MIG][15.0] openupgrade_framework

legalsylvain committed:
[REF] adapt framework, due to renaming handle_view_error into _raise_view_error

legalsylvain committed:
fixup! [WIP][15.0] base module

legalsylvain committed:
fixup! fixup! [WIP][15.0] base module

StefanRijnhart committed:
[OU-FIX] Migration scripts were not run during install

Fixes: #3285

jcdrubay committed:
Update method _postprocess_view of ir_ui_view by removing the input arg validate as it is not anymore in odoo 15

jcdrubay committed:
[MIG] remove the odoo_patch on function _postprocess_view which became useless after the input argument validate has been removed

jcdrubay committed:
Adjustements for code style

royle-vietnam committed:
[IMP] base: override to not raise an exception

Because we captured the exception in _raise_view_error and archived that
view, so info is None, but it is called to info.get('select') in
NameManager.check, which will raise an exception AttributeError, so we
need to override to not raise an exception

OCA-git-bot committed:
[UPD] README.rst

royle-vietnam committed:
[IMP] openupgrade_framework: tracking number of rows per table before
and after the migration

MiquelRForgeFlow committed:
[IMP] openupgrade_framework: add environ parameter to force demo

StefanRijnhart committed:
Revert "[IMP][15.0] openupgrade_framework: tracking number of rows per table before and after the migration"

royle-vietnam committed:
[IMP] openupgrade_framework: to not crash, simply not deactivating the
views

OCA-git-bot committed:
[UPD] README.rst

pedrobaeza committed:
[FIX] openupgrade_framework: Proper module dependency resolution

MiquelRForgeFlow committed:
[FIX] framework: monkeypatch needs to patch original method

MiquelRForgeFlow committed:
[FIX] framework: monkeypatch needs to patch original method (fixup)

hbrunn committed:
[MIG] base v16

hbrunn committed:
[MIG] openupgrade_framework v16

legalsylvain commented:
/ocabot migration base

legalsylvain commented:
/ocabot migration base

legalsylvain commented:
/ocabot migration base

StefanRijnhart commented on openupgrade_scripts/scripts/base/16.0.1.3/post-migration.py:
```diff
@@ -0,0 +1,45 @@
+    ) and not openupgrade.table_exists(env.cr, "_ir_translation"):
+        openupgrade.rename_tables(env.cr, [("ir_translation", "_ir_translation")])
+
+    openupgrade_160.migrate_translations_to_jsonb(
```
Rather than on a module by module basis, can we do this here for all translatable fields in the database? I see the lib method uses the models from the environment, so that would have to be addressed.

hbrunn replied:
this would mean we have to convert columns to jsonb manually, and we'd also not be able to use https://github.com/OCA/OCB/blob/16.0/odoo/tools/translate.py#L1551 because this also expects the registry to be loaded for the field in question.

Or do you mean to move this to end-migration and do it in bulk there?

pedrobaeza replied:
The second one as we both talked in OCA days :wink:

StefanRijnhart replied:
Well, if it was the end-migration the process would not be applied to modules that are not loaded at that moment, such as when a setup is migrated from 15 to 17 and some modules are not provided for 16.0. So I would prefer here using our own SQL logic (also for performance).

StefanRijnhart replied:
I mean, you can still do it in the end-migration so that you can benefit from the models that *are* in the environment to report any non-standard table names, but I would prefer to have models that are not loaded at that point covered as well, as much as possible.

pedrobaeza replied:
OK, I see, but I'm not sure if we can deduce which SQL columns to convert.

StefanRijnhart replied:
I think we can go by the value in the `translate` column in `ir_model_fields`.

StefanRijnhart replied:
Oh, I must admit that an exception are the fields with a `callable` translate value. This is not available for fields that are not loaded in the environment (at the end stage).
Fortunately, this is not entirely ambiguous because those fields' database translations have type `model_terms` rather than `model`.

hbrunn replied:
I've implemented that in https://github.com/OCA/OpenUpgrade/pull/3680/commits/25830ead71b0183c91c7e486ad9295bcca95a6b2

StefanRijnhart replied:
Brilliant, thanks! Shall I recreate a new version of the template database and install an additional language in it? We can then add a test :heart_eyes:
Do you plan a special treatment in the end-migration for fields that have a callable value for `translate` (once the field is loaded in the environment)?

hbrunn replied:
I actually didn't think about translate=callable yet, I think that can be a later problem.

The DB with an additional language would be very welcome indeed.

StefanRijnhart commented:
Please copy any preexisting `company_registry` values on `res.company` records to `res.partner`.

hbrunn committed:
[ADD] implement tests of migrations

hbrunn commented:
I've added the testing part in https://github.com/OCA/OpenUpgrade/pull/3680/commits/3ea5741f512575782df1ca2558f5452cfe2f588f, but the [test for base I added](https://github.com/OCA/OpenUpgrade/pull/3680/commits/3ea5741f512575782df1ca2558f5452cfe2f588f#diff-08e429ab7266e360a22d3570870434ee1a74fa2c06f1bedacd5f455e67796454R7) will fail on CI because because the data isn't set anywhere.

In older times we had the tests load a bunch of yaml files via `--test-file`, but that's only python by now and then also unittests only.

What do you think of just executing all sql in `openupgrade_scripts/scripts/$module/$version/tests/*.sql` before `$module`'s pre-migration script is run?

In the case of the base migration, that would be something like

```sql
--- set company_registry in v15 to be sure we migrate it correctly to v16
update res_company c set company_registry='424242' from ir_model_data d where d.name='main_company' and c.id=d.res_id
```

pedrobaeza commented:
OK, seems a good option.

hbrunn commented:
test data now happens in https://github.com/OCA/OpenUpgrade/pull/3680/commits/fdfb642b31a4f10e242afb277c347196a356525e

StefanRijnhart commented:
Thanks for reimplementing tests! On the one hand, this is a really lean and clean solution, but on the other hand, if we would be able to create data through one or more Odoo shell scripts, that might make it easier for contributors to create (more complex) tests. That does imply installing the previous version of Odoo first in the Github workflow first (like we had in the past) to run the shell scripts on the test database. How does everybody feel about that?

StefanRijnhart commented on openupgrade_scripts/scripts/base/16.0.1.3/pre-migration.py:
```diff
@@ -0,0 +1,92 @@
+        "JOIN ir_model m ON f.model_id=m.id WHERE f.translate"
+    )
+    for field, model in cr.fetchall():
+        table = model.replace(".", "_")
```
I think we will end up here with a hard coded list of non-standard table names here to cover those models as well (if only to get rid of the warnings). Maybe you can already introduce it at this point and populate it with the known ir.actions* mapping.

hbrunn commented:
So we do something like

```shell
for f in $script_path/*/*/tests/data*.py; do
    odoo-bin(N-1) shell < $f
```

in the test script?

And do we want to keep SQL as an alternative or do everything this way?

StefanRijnhart commented:
Yes, that's it! And keeping the SQL method apart from that would be redundant I feel.

StefanRijnhart commented:
@hbrunn new database is now uploaded with the French language installed.
```
15.0=> select count(*), lang from ir_translation group by lang;
 count | lang  
-------+-------
 56087 | fr_FR
(1 row)
```
I reran tests and they are still green! Would you like to add another test specifically to test one or two translations?

hbrunn commented:
I would, but probably won't this week. Thank you for this, I clicked 'rerun all jobs'

StefanRijnhart commented:
No problem. Would you also be so kind as to add an annotated analysis_work file with a TODO for the fields that have a callable `translate` attribute?

pedrobaeza committed:
[OU-IMP] base: Spanish address format

After odoo/odoo#110918, we put this new format on migrated v15 DBs
for those not containing the new format as they were populated in v14
before the commit.

pedrobaeza commented:
I think #3706 can be also included here, as the v15 DB may be populated before this Odoo commit was present.

TadeuszKarpinski commented:
How can I start tests locally? Init 15.0 db with data, then migrate and run tests?

hbrunn commented:
@TadeuszKarpinski it's much too early to run any user tests, this PR is the bare necessities to start writing migration scripts

hbrunn commented:
@StefanRijnhart added a test for french and the _work file, @pedrobaeza added your commit

hbrunn committed:
[UPD] isort for pre-commit

pedrobaeza commented on openupgrade_framework/README.rst:
```diff
@@ -14,13 +14,16 @@ Openupgrade Framework
-    :target: https://translation.odoo-community.org/projects/OpenUpgrade-15-0/OpenUpgrade-15-0-openupgrade_framework
+    :target: https://translation.odoo-community.org/projects/OpenUpgrade-16-0/OpenUpgrade-16-0-openupgrade_framework
     :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runboat-Try%20me-875A7B.png
```
There's no such runboat, and it will be misleading anyway thinking that users can try the migration.

pedrobaeza commented on docsource/modules150-160.rst:
```diff
@@ -66,7 +66,7 @@ Module coverage 15.0 -> 16.0
 | barcodes_gs1_nomenclature                       |                      |                                                 |
 +-------------------------------------------------+----------------------+-------------------------------------------------+
-| base                                            |                      |                                                 |
+| base                                            |                      |Partial                                          |
```
The third column is for indicating if no DB layout changes. It should be filled in the second column, and `Partial` is not recognized for the modules to test according the "parser" integrated in the GH action. They are `Done` or `Nothing to do`.

You can put `Done. Partial` though.

pedrobaeza commented on openupgrade_scripts/scripts/base/16.0.1.3/post-migration.py:
```diff
@@ -0,0 +1,14 @@
+@openupgrade.migrate()
+def migrate(env, version):
+    openupgrade.load_data(env.cr, "base", "16.0.1.3/noupdate_changes.xml")
+
```
Please don't put empty lines inside a method.

pedrobaeza commented on openupgrade_scripts/scripts/base/16.0.1.3/pre-migration.py:
```diff
@@ -0,0 +1,114 @@
+_logger = logging.getLogger(__name__)
+
+try:
+    from odoo.addons.openupgrade_scripts.apriori import merged_modules, renamed_modules
```
Why not letting it failed directly? If not, people don't read...

hbrunn replied:
I just took that over from the previous version, remove then?

pedrobaeza replied:
I think it's better to crash if you have this situation than having a weird error and a crash later when running the scripts.

pedrobaeza commented on openupgrade_scripts/scripts/base/16.0.1.3/pre-migration.py:
```diff
@@ -0,0 +1,114 @@
+            "recommended to run the Odoo with --load=openupgrade_framework "
+            "when migrating your database."
+        )
+
```
Please avoid empty lines inside a method. It's recommended in PEP8 and visually you already have the comment lines for separating blocks.

hbrunn commented:
coverage would be nice though, but that can be another PR

pedrobaeza commented:
And please squash commits for merging this.

hbrunn commented:
rebased

StefanRijnhart commented:
Thanks for the updates, and great work overall!
